Bugfix/catch reindex exception

richaber opened the pull request:
Closes #109, #175

This PR catches Throwables thrown by Algolia\AlgoliaSearch\SearchIndex::saveObjects as noted in #109 and sends their messages to error_log(), like so...
```
[31-Aug-2021 23:42:29 UTC] Record at the position 5 objectID=1196-0 is too big size=1306368/10000 bytes. Please have a look at https://www.algolia.com/doc/guides/sending-and-managing-data/prepare-your-data/in-depth/index-and-records-size-and-usage-limitations/#record-size-limits
```

This won't _prevent_ a record that is too large from being rejected, but it should at least make debugging a little easier.

This PR also sets a flag in `Algolia_Admin::re_index()` during reindexing, to prevent duplicate API calls in `Algolia_Admin::update_records()`.

jrfoell commented:
Rather than skipping the request (per your suggestion), I like the try/catch method that you're implementing as part of this PR. If someone (in the future) specifically wants to skip large records because they're taking too long, I would suggest creating a filter to allow them to conditionally skip for their particular install.

richaber commented:
Updated to skip saveObjects if _wp_json_sanity_check throws an Exception, since that is our "sanitization."

Since re-indexing is called from a WP-CLI command, or on an admin-ajax request in Dashboard, I could probably collect all the Throwables during reindexing, and use them as a message for site administrator via wp-cli warning, ajax response alert, or maybe even "yet another admin notice™", but I'll likely save that for another day.

For now, I think this is likely "good enough" as a developer should be able to turn on `WP_DEBUG_LOG` and see the Exception messages in the log if they're seeing unexpected results.

Testing appears good.

During re-indexing, I do not end up in a state where the index is mangled by an uncaught BadRequestException, the offending record just doesn't make it to the API, and subsequent inoffensive records are processed as expected. 

Also added in a fix for #175. It's not the most elegant solution, but it's simple and appears to get the job done.

richaber commented:
Making a note to self that as of this morning, my testing index is completely empty, and I am not currently certain if this branch is responsible or not.

richaber commented:
Added two new filters `algolia_update_records` and `algolia_re_index_records`, per @jrfoell's suggestion.

This would allow someone to check their record sizes with something like this during a re-index...
```
/**
 * Filters the records to be reindexed.
 *
 * @param array  $records  Array of records to re-index.
 * @param int    $page     Page to re-index.
 * @param string $index_id The index ID without prefix.
 *
 * @return mixed
 */
function my_algolia_re_index_records( $records, $page, $index_id ) {

	foreach ( $records as $key => $record ) {
		$size = strlen( wp_json_encode( $record ) );
		if ( $size > 10000 ) {
			// phpcs:ignore -- Need a real logger.
			error_log(
				sprintf(
					'Reindexing Records objectID=%1$s appears to be too large. Size %2$d / 10000 bytes.',
					$record['objectID'],
					$size
				)
			);
			unset( $records[ $key ] );
		}
	}

	return $records;
}

add_filter( 'algolia_re_index_records', 'my_algolia_re_index_records', 10, 3 );
```
Which would log a message like "Reindexing Records objectID=1196-0 appears to be too large. Size 1306404 / 10000 bytes." in the error log.

And with something like this during an update...
```
/**
 * Filters the records to be updated.
 *
 * @param array  $records  Array of records to update.
 * @param object $item     The item to update records for.
 * @param string $index_id The index ID without prefix.
 */
function my_algolia_update_records( $records, $item, $index_id ) {

	foreach ( $records as $key => $record ) {
		$size = strlen( wp_json_encode( $record ) );
		if ( $size > 10000 ) {
			// phpcs:ignore -- Need a real logger.
			error_log(
				sprintf(
					'Updating Records objectID=%1$s appears to be too large. Size %2$d / 10000 bytes.',
					$record['objectID'],
					$size
				)
			);
			unset( $records[ $key ] );
		}
	}

	return $records;
}

add_filter( 'algolia_update_records', 'my_algolia_update_records', 10, 3 );
```
Which would log a message like "Updating Records objectID=1196-0 appears to be too large. Size 1306404 / 10000 bytes." in the error log.

jrfoell commented on includes/indices/class-algolia-index.php:
```diff
@@ -322,14 +332,55 @@ abstract protected function get_records( $item );
+		 * @param object $item     The item to update records for.
+		 * @param string $index_id The index ID without prefix.
+		 */
+		$records = apply_filters(
+			'algolia_update_records',
+			$records,
+			$item,
+			$this->get_id()
+		);
```
Nice job :muscle:

jrfoell commented on includes/indices/class-algolia-index.php:
```diff
@@ -410,13 +468,47 @@ public function re_index( $page ) {
+			 * @param int    $page     Page to re-index.
+			 * @param string $index_id The index ID without prefix.
+			 */
+			$records = apply_filters(
+				'algolia_re_index_records',
+				$records,
+				$page,
+				$this->get_id()
+			);
```
Same here :muscle:

jrfoell reviewed (approved):
Nice work!
